pkg/processor/trigger/trigger.go

```diff
@@ -424,7 +431,101 @@ func (at *AbstractTrigger) prepareEvent(event nuclio.Event, workerInstance *work
 	}
 
 	// Not a cloud event
-	event.SetID(nuclio.ID(uuid.New().String()))
+	// id won't be empty only in batching mode
+	if event.GetID() == "" {
+		event.SetID(nuclio.ID(uuid.New().String()))
+	}
 	event.SetTriggerInfoProvider(at)
 	return event, nil
 }
+
+func (at *AbstractTrigger) StartBatcher(batchTimeout time.Duration, workerAvailabilityTimeout time.Duration) {
+	for {
+		batch, responseChans := at.Batcher.WaitForBatch(batchTimeout)
+		// allocate a worker
+		workerInstance, err := at.WorkerAllocator.Allocate(workerAvailabilityTimeout)
+		if err != nil {
+			at.UpdateStatistics(false, uint64(len(batch)))
+			workerError := errors.Wrap(err, "Failed to allocate worker")
+			for _, channel := range responseChans {
+				go channel.Write(at.Logger, &runtime.ResponseWithErrors{SubmitError: workerError})
+			}
+			return
+		}
+
+		// submit batch to the worker
+		at.SubmitBatchAndSendResponses(batch, responseChans, workerInstance)
+
+		// release worker when we're done
+		at.WorkerAllocator.Release(workerInstance)
+	}
+}
```

Comment:
Maybe instead of having this method and the one in http trigger overriding it. How about:
```suggestion
func (at *AbstractTrigger) StartBatcher(batchTimeout time.Duration, workerAvailabilityTimeout time.Duration) {
	for {
		batch, responseChans := at.Batcher.WaitForBatch(batchTimeout)
		// allocate a worker
		workerInstance, err := at.WorkerAllocator.Allocate(workerAvailabilityTimeout)
		if err != nil {
			at.UpdateStatistics(false, uint64(len(batch)))
			workerError := errors.Wrap(err, "Failed to allocate worker")
			for _, channel := range responseChans {
				go channel.Write(at.Logger, &runtime.ResponseWithErrors{SubmitError: workerError})
			}
			return
		}
		
		at.preBatchHooks()

		// submit batch to the worker
		at.SubmitBatchAndSendResponses(batch, responseChans, workerInstance)

		// release worker when we're done
		at.WorkerAllocator.Release(workerInstance)
		
		at.postBatchHooks()
	}
}
```
Then instead of implementing the whole method, https trigger can implement only `preBatchHooks` and `postBatchHooks`

Reply:
@quaark great idea!